examples/device/c/device.c

```diff
@@ -426,92 +445,254 @@ int QDMI_control_get_data_dev(QDMI_Job job, const QDMI_Job_Result result,
+        *size_ret = req_size;
+      }
+      if (data != NULL) {
+        if (size < req_size) {
+          free((void *)raw_data_split);
+          free(raw_data);
+          return QDMI_ERROR_INVALIDARGUMENT;
+        }
+
+        char *data_ptr = data;
+        strcpy(data_ptr, raw_data_split[0]);
+        data_ptr += num_qubits;
+        for (int j = 1; j < job->num_shots; ++j) {
+          if (strcmp(raw_data_split[j], raw_data_split[j - 1]) != 0) {
+            *data_ptr++ = ',';
+            strcpy(data_ptr, raw_data_split[j]);
+            data_ptr += num_qubits;
+          }
+        }
+        *data_ptr++ = '\0';
+      }
+    } else {
+      const int req_size = count * (int)sizeof(int);
+      if (size_ret != NULL) {
+        *size_ret = req_size;
+      }
+      if (data != NULL) {
+        if (size < req_size) {
+          free((void *)raw_data_split);
+          free(raw_data);
+          return QDMI_ERROR_INVALIDARGUMENT;
+        }
+
+        int *data_ptr = data;
+        int n = 1;
+        for (int j = 1; j < job->num_shots; ++j) {
+          if (strcmp(raw_data_split[j], raw_data_split[j - 1]) != 0) {
+            *data_ptr++ = n;
+            n = 1;
+          } else {
+            ++n;
+          }
+        }
+        *data_ptr++ = n;
+      }
+    }
+    free((void *)raw_data_split);
+    free(raw_data);
+    return QDMI_SUCCESS;
+  }
+  if (result == QDMI_JOB_RESULT_STATEVECTOR_DENSE) {
     int num_qubits = 0;
     QDMI_query_device_property_dev(QDMI_DEVICE_PROPERTY_QUBITSNUM, sizeof(int),
                                    &num_qubits, NULL);
     if (data != NULL) {
-      if (size < count * (num_qubits + 1)) {
-        free((void *)raw_data_split);
-        free(raw_data);
+      if (size < (2ULL << num_qubits) * sizeof(double)) {
         return QDMI_ERROR_INVALIDARGUMENT;
       }
-      strcpy((char *)data, raw_data_split[0]);
-      data += num_qubits;
-      if (count > 1) {
-        *(char *)data = ',';
+      memcpy(data, job->state_vec, (2ULL << num_qubits) * sizeof(double));
+    }
+    if ((size_ret) != NULL) {
+      *(size_ret) = (int)((2ULL << num_qubits) * sizeof(double));
+    }
+    return QDMI_SUCCESS;
+  }
+  if (result == QDMI_JOB_RESULT_STATEVECTOR_SPARSE_KEYS) {
+    int dense_size = 0;
+    QDMI_control_get_data_dev(job, QDMI_JOB_RESULT_STATEVECTOR_DENSE, 0, NULL,
+                              &dense_size);
+    double *dense_data = malloc(dense_size);
+    QDMI_control_get_data_dev(job, QDMI_JOB_RESULT_STATEVECTOR_DENSE,
+                              dense_size, dense_data, NULL);
+    int num_qubits = 0;
+    QDMI_query_device_property_dev(QDMI_DEVICE_PROPERTY_QUBITSNUM, sizeof(int),
+                                   &num_qubits, NULL);
+    // count non-zero elements
+    int count = 0;
+    for (size_t i = 0; i < 1ULL << num_qubits; i++) {
+      if (dense_data[2 * i] != 0.0 || dense_data[(2 * i) + 1] != 0.0) {
+        count++;
+      }
+    }
+    if (data != NULL) {
```

Comment:
The sparse functions for the state vector and the probability vector still feature the kind of code duplication that was also present in the histogram output before this PR.
Would be great if we could also reduce the duplication here.

Another important point in that regard: in most of these routines, the call to the `DENSE` variant can be omitted and the `job->state_vec` data can be directly used.

This comment holds for both the C and the C++ device.

Reply:
Is that desired to refer to `job->state_vec` and respective data structures rather than calling the `DENSE` function? Calling the dense function is more generic and also works if one decides to store the results somewhere or somehow different.

Reply:
I believe that it is fine to use whatever the job stores as it's results. This is up to the device implementer anyway. We have to keep in mind that not all the devices will be able to provide all of the results for a given job.
This will most likely require a way to query, which kind of results are supported by the device and then explicitly requesting them as part of the job (parameters).
But I guess this is something for a different PR.

I'd generally try to keep the example device implementations as simple as possible.

Reply:
Another argument: I wouldn't like to set the precedent that it is common practice to call the dense methods as part of the sparse methods. The sparse methods are there for a reason: in many practical cases you won't have the memory in all the world to store the dense result, but you might be able to directly report back the sparse result.
Us even being able to implement it in this fashion is simply due to the fact that we are only looking at a small example use case and that we generate a random result in the form of a random dense state-vector.

I would expect that practical implementations of simulators actually would implement this the other way round in many cases. That is, they offer the sparse result and, optionally, a way to create a dense result from that assuming that enough memory is available.

Reply:
True, agreed.

Reply:
https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/82/commits/ab65825961ba8e5b7932d96b71692151754550b2 now contains these refactors and considerably shrinks the implementation. I think it looks quite nice now.
Given all tests pass, I'd be up for merging this.